velox/docs/develop/aggregate-functions.rst

```diff
@@ -109,24 +109,323 @@ accordingly.  Data is stored in the following order:
+and velox/exec/tests/SimpleArrayAggAggregate.cpp.
+
+A simple aggregation function is implemented as a class as the following.
+
+.. code-block:: c++
+
+  // array_agg(T) -> array(T) -> array(T)
+  class ArrayAggAggregate {
+   public:
+    // Type(s) of input vector(s) wrapped in Row.
+    using InputType = Row<Generic<T1>>;
+    using IntermediateType = Array<Generic<T1>>;
+    using OutputType = Array<Generic<T1>>;
+
+    // Optional. Default is true.
+    static constexpr bool default_null_behavior_ = false;
+
+    // Optional.
+    static bool toIntermediate(
+      exec::out_type<Array<Generic<T1>>>& out,
+      exec::optional_arg_type<Generic<T1>> in);
+
+    struct AccumulatorType { ... };
+  };
+
+The author declares the function's input type, intermediate type, and output
+type in the simple aggregation function class. The input type must be the
+function's argument type(s) wrapped in a Row<> even if the function only takes
+one argument. This is needed for the SimpleAggregateAdapter to parse input
+types for arbitrary aggregation functions properly.
+
+The author can define an optional flag `default_null_behavior_` indicating
+whether the aggregation function has default-null behavior. This flag is true
+by default. Next, the class can have an optional method `toIntermediate()`
+that converts the aggregation function's raw input directly to its intermediate
+states. Finally, the author must define a struct named `AccumulatorType` in
+the aggregation function class. We explain each part in more details below.
+
+Default-Null Behavior
+^^^^^^^^^^^^^^^^^^^^^
+
+When adding raw inputs or intermediate states to accumulators, aggregation
+functions of default-null behavior ignore the input values that are nulls. For
+raw inputs that consist of multiple columns, an entire row is ignored if at
+least one column is null at this row. Below is an example.
+
+.. code-block:: sql
+
+  SELECT sum(c0) FROM (values (cast(NULL as bigint), 10), (NULL, 20), (NULL, 30)) AS t(c0, c1); -- NULL
+
+When generating intermediate or final output results from accumulators,
+aggregation functions of default-null behavior produce nulls for groups of no
+input row or only null rows. Another example is given below.
+
+.. code-block:: sql
+
+  SELECT sum(c0) FROM (values (1, 10), (2, 20), (3, 30)) AS t(c0, c1) WHERE c1 > 40; -- NULL
+
+Most aggregation functions have default-null behavior. An example is in
+SimpleAverageAggregate.cpp. On the other hand, SimpleArrayAggAggregate.cpp has
+an example of non-default-null behavior.
+
+This flag affects the C++ function signatures of `toIntermediate()` and methods
+in the `AccumulatorType` struct.
+
+toIntermediate
+^^^^^^^^^^^^^^
+
+The author can optionally define a static method `toIntermediate()` that
+converts a raw input to an intermediate state. If defined, this function is
+used in `abandonedPartialAggregation`. If the aggregaiton function has
```

Comment:
What is 'abandonedPartialAggregation' ?

Reply:
That is the query plan that calls toIntermediate() instead of doing partial aggregation step. How about I change this sentence to "If defined, this function is used in query plans that abandon the partial aggregation step"?

Reply:
Thanks @kagamiori That sounds fine.